src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java

```diff
@@ -190,44 +197,38 @@ public static void scan(Run build, PrintStream buildLog) {
      */
     private static List<FoundFailureCause> findCauses(final Collection<FailureCause> causes,
                                                       final Run build, final PrintStream buildLog) {
-        final List<FoundFailureCause> foundFailureCauseList =
-                Collections.synchronizedList(new LinkedList<FoundFailureCause>());
+        final List<FoundFailureCause> foundFailureCauseList = new ArrayList<FoundFailureCause>();
         long start = System.currentTimeMillis();
-        ThreadPoolExecutor executor = (ThreadPoolExecutor)Executors.
-                newFixedThreadPool(PluginImpl.getInstance().getNrOfScanThreads());
+
+        THREAD_POOL_EXECUTOR.setCorePoolSize(PluginImpl.getInstance().getNrOfScanThreads());
+        THREAD_POOL_EXECUTOR.setMaximumPoolSize(PluginImpl.getInstance().getNrOfScanThreads());
+
         buildLog.println("[BFA] Scanning build for known causes...");
-        for (final FailureCause cause : causes) {
-            executor.execute(new Runnable() {
+
+        Runnable runnable = new Runnable() {
                 @Override
                 public void run() {
-                    Thread.currentThread().setName("BFA-scanner-"
-                                                           + build.getFullDisplayName() + ": "
-                                                           + cause.getName());
-                    List<FoundIndication> foundIndications = findIndications(cause, build, buildLog);
-                    if (!foundIndications.isEmpty()) {
-                        FoundFailureCause foundFailureCause = new FoundFailureCause(cause, foundIndications);
-                        foundFailureCauseList.add(foundFailureCause);
-                    }
+                    Thread.currentThread().setName("BFA-scanner-" + build.getFullDisplayName());
+                    foundFailureCauseList.addAll(findIndications(causes, build, buildLog));
```

Comment:
It could create delay at first time, but after that user can increase number of threads if server has enough capacity or decrease number of regular expressions (or optimize it :) ).